### PR TITLE
MySQL support for UPDATE/DELETE to use ORDER and LIMIT

### LINKED ISSUE #9195: Query::update() does not support LIMIT
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.2.12
* Platform and Target: Apache2, MySQL, PHP 7

### What you did

Trying to create a custom update query that has a LIMIT 1

```
    $query = $documents->query()->update()->set(['title'=>'foo'])->limit(1);
    Log::info((string)$query);
```

Output:

```
   UPDATE `documents` SET `title` = :c0
```

### Expected Behavior

MySQL supports LIMIT on UPDATE queries, but the ORM is not adding the LIMIT clause.

### Actual Behavior

All records are updated. :cry:

### PR DESCRIPTION
**Pull request by thinkingmedia**
fixes #9195 

Adds support for more clauses with MySQL for the UPDATE and DELETE queries. Specifically adds support for ORDER and LIMIT.

**TODO**:

[ ] Postgres support for ORDER in UPDATE with unit tests.
[ ] Postgres support for LIMIT in UPDATE with unit tests.
[ ] Postgres support for ORDER in DELETE with unit tests.
[ ] Postgres support for LIMIT in DELETE with unit tests.

[ ] SQLite support for ORDER in UPDATE with unit tests.
[ ] SQLite support for LIMIT in UPDATE with unit tests.
[ ] SQLite support for ORDER in DELETE with unit tests.
[ ] SQLite support for LIMIT in DELETE with unit tests.

[ ] SQL Server support for ORDER in UPDATE with unit tests.
[ ] SQL Server support for LIMIT in UPDATE with unit tests.
[ ] SQL Server support for ORDER in DELETE with unit tests.
[ ] SQL Server support for LIMIT in DELETE with unit tests.

[ ] MySQL support for ORDER in UPDATE with unit tests.
[ ] MySQL support for LIMIT in UPDATE with unit tests.
[ ] MySQL support for ORDER in DELETE with unit tests.
[ ] MySQL support for LIMIT in DELETE with unit tests.

### CODE REVIEW

**Comment by thinkingmedia** [comment]
I'm not sure why modifiers broke. I'm investigating.

**Comment by markstory** [comment]
If a plugin author would to use these clauses, and the application used a postgres/sqlite/sqlserver database wouldn't the behavior not be correct? Right now the behavior is consistently 'bad'.

**Comment by thinkingmedia** [comment]
@markstory yes. Should this be done on the base QueryCompiler?

I guess I could write unit tests for postgress/sqlite/sqlserver and test them on travis.

**Comment by markstory** [comment]
@thinkingmedia I wonder if this wouldn't be better done in the existing QueryCompiler, and let the databases that don't support these features emit SQL syntax errors when applications/plugins are trying to use features that don't exist.

**Comment by lorenzo** [comment]
This can be easily emulated in all drivers. If we are adding support for it, I think we should strive for offering an uniform experience.

**Comment by markstory** [comment]
@lorenzo It looks like both SQLServer and Postgres will need emulation. The examples I've seen recommend using subqueries or common table expressions.

**Comment by thinkingmedia** [comment]
# Postgres

- does not support ORDER/LIMIT in UPDATE
- does not support ORDER/LIMIT in DELETE

> Postgres has a horrid manual by the way.

# SQLite

- does not support ORDER/LIMIT in UPDATE
- does not support ORDER/LIMIT in DELETE

# SQL Server

- does not support LIMIT but has TOP variant in UPDATE
- does not support ORDER BY in UPDATE
- DELETE same as UPDATE

----

So, as you can see this is not a generic feature. 

> This can be easily emulated in all drivers. If we are adding support for it, I think we should strive for offering an uniform experience.

In a way, I agree and disagree. For example, the `Table::deleteAll()` should work consistently across all drivers. We could add extra options for order and limit.

The query builder should remain true to SQL. For example;

```
   $table->query()->delete()->where(...)->order(..)->limit(...);
```

The only way to get the above to work across all drivers is if a sub-query were injected into the SQL. That creates this "magical" modification to the above query. I don't think we should do that. The query builder should be used to only create SQL queries as defined by the programmer.

With that said, the `deleteAll` and `updateAll` methods of the table should be consistent across all drivers.

@markstory if we add support for this to the QueryCompiler we introduce invalid SQL syntax for 3 out of 4 of the drivers. With the exception to SQL server, but that uses it's own compiler anyway. So the query compiler would be updated just for MySQL, and Postgres and SQLite would fail.

It makes more sense to me to have a specific compiler for MySQL that adds support, and makes the support look intentional.

# New Ideas

Given everything I just said.

We could add this to the base QueryCompiler and then I can add a `_buildLimitPart` method to the `SqliteCompiler` that raises an exception. I can repeat this in the other compilers if the clause is not supported.

This allows the ORDER/LIMIT clauses to be injected into the SQL if new drivers are added by the programmer (on their own), but for the ORMs we know do not support it we can do an early exception on that.

**Comment by markstory** [comment]
>  That creates this "magical" modification to the above query. I don't think we should do that. The query builder should be used to only create SQL queries as defined by the programmer.

We do this quite a bit already. Both the Postgres dialect and SQLServer dialects rewrite queries that could not otherwise exist on those SQL dialects.

> we introduce invalid SQL syntax for 3 out of 4 of the drivers. With the exception to SQL server, but that uses it's own compiler anyway. So the query compiler would be updated just for MySQL, and Postgres and SQLite would fail.

Yes, but we avoid problems where a plugin author is dependent on `UPDATE ... LIMIT` and the application is running with postgres. With the invalid SQL approach the code can't pretend to work, but silently do the wrong thing.

**Comment by lorenzo** [comment]
@markstory yes, that's why I said it can be *emulated* easily :P

**Comment by thinkingmedia** [comment]
@markstory @lorenzo where is this rewriting happening? Is it done in the query compilers?

**Comment by lorenzo** [comment]
@thinkingmedia usually in the query translators. But it can also happen in the compilers

**Comment by thinkingmedia** [comment]
@lorenzo ah, I found `SqlserverDialectTrait::_pagingSubquery`. Is this an example of what you mean?

**Comment by markstory** [comment]
@thinkingmedia Yes, also `PosgresDialect::_transformFunctionExpression()` and `PostgresDialect::_insertQueryTranslator()`.

**Comment by thinkingmedia** [comment]
I'm researching how to emulate a LIMIT clause on an UPDATE in Postgres, and while I've found many answers on google. It's very difficult to rewrite the Query object in `PostgresDialectTrait` without knowing the programmers intentions.

In Postgres you can use the FROM clause in an UPDATE. So I was going to rewrite the query to use that with a SELECT LIMIT. The problem is that I can't get the name of the primary keys to use in a WHERE clause.

For example:

```
UPDATE articles SET title = 'foo' LIMIT 5;
```

Would be rewritten 

```
UPDATE articles SET title = 'foo' FROM (SELECT id FROM articles LIMIT 5) __cake__ WHERE id = __cake__.id;
```

I don't think it's possible to rewrite the above via the QueryCompiler.

Furthermore, this just **feels wrong**. I think Query should support FROM in UPDATES just for Postgres, and it's up to the developer t know how to implement a query for a given driver. There is more at risk than just general SQL support. There are performance and concurrency issues for the different drivers.

Anyone care to share their thoughts?

**Comment by dakota** [comment]
I'm strongly of the opinion that the ORM/Query builder should abstract away database specific implementations as much as possible. If something cannot be presented to the user with the same interface across any and all database engines, then the framework should not offer that feature.

The framework should however give the developer  an interface to manipulate parts of the query if needed.

**Comment by lorenzo** [comment]
@thinkingmedia I can give you a hand, I think I know how to do this. Will push a few commits tomorrow or the day after

**Comment by thinkingmedia** [comment]
@lorenzo thank you. That would be great. I'm not a Postgres user so I'm not aware of all the little tricks to do things.

@dakota you are missing part of the discussion where the ORM silently executes a query without the supported clauses. Which is what originated this PR. As it stands now, the ORM will create an update query with an order and limit clause and execute it against a Postgres driver without any error or warning. The resulting SQL is not what the programmer intended and the results are unpredictable. What's happen in the PR now is an attempt to handle update/limit across all drivers, and if that can not be achieved an notice has to be give that the query contains an unsupported clause. So we try to get support on all drivers first.

Later tonight I'll take a look at Sqlite.

**Comment by markstory** [comment]
Good point about the `WHERE` less update. I don't think that is ever a 'safe' query, but it a good example of why rewriting queries may be feasible. If we can safely emulate features in databases that don't support them I think we should. But measuring safety can be tricky.

File: src/Database/Dialect/PostgresDialectTrait.php [hunk 1]
```diff
@@ -62,6 +65,69 @@ protected function _transformDistinct($query)
...
+    {
+        $limit = $query->clause('limit');
+
+        if ($limit) {
+            $from = (new Query($query->connection()))
+                ->select('*')
+                ->from($query->clause('update'))
+                ->where($query->clause('where'))
+                ->limit($limit)
+                ->epilog('FOR UPDATE');
```
Comment (markstory) [inline src/Database/Dialect/PostgresDialectTrait.php]: :+1:

File: src/Database/Dialect/PostgresDialectTrait.php [hunk 2]
```diff
@@ -62,6 +65,69 @@ protected function _transformDistinct($query)
...
+
+            $outer = clone $query;
+            $outer
+                ->from(['__cake_paging__' => $from], true)
```
Comment (markstory) [inline src/Database/Dialect/PostgresDialectTrait.php]: Its not a paging query, perhaps `__cake_update__`?

File: src/Database/Dialect/PostgresDialectTrait.php [hunk 3]
```diff
@@ -62,6 +65,69 @@ protected function _transformDistinct($query)
...
+     * @param int $offset The number of rows to offset.
+     * @return \Cake\Database\Query Modified query object.
+     */
+    protected function _limitSubquery($original, $limit, $offset) {
+        $field = '_cake_paging_._cake_page_rownum_';
+        $order = $original->clause('order') ?: new OrderByExpression('(SELECT NULL)');
+
+        $query = clone $original;
+        $query->select([
+            '_cake_page_rownum_' => new UnaryExpression('ROW_NUMBER() OVER', $order)
```
Comment (markstory) [inline src/Database/Dialect/PostgresDialectTrait.php]: This looks like SQLServer code.

File: src/Database/Driver/Mysql.php [hunk 4]
```diff
@@ -159,6 +160,14 @@ public function supportsDynamicConstraints()
     }
 
     /**
+     * {@inheritDoc}
+     */
+    public function newCompiler()
+    {
+        return new MysqlCompiler();
```
Comment (markstory) [inline src/Database/Driver/Mysql.php]: This isn't in the pull request.

File: src/Database/Dialect/PostgresDialectTrait.php (hunk 1 above)
Comment (lorenzo) [inline src/Database/Dialect/PostgresDialectTrait.php]: I don't think there is a need for this, all queries are atomic, even if you use subqueries.

File: src/Database/Dialect/PostgresDialectTrait.php [hunk 5]
```diff
@@ -62,6 +65,69 @@ protected function _transformDistinct($query)
...
+        if ($offset) {
+            $outer->where(["$field > " . (int)$offset]);
+        }
+        if ($limit) {
+            $value = (int)$offset + (int)$limit;
+            $outer->where(["$field <= $value"]);
+        }
+
+        return $outer;
+    }
```
Comment (lorenzo) [inline src/Database/Dialect/PostgresDialectTrait.php]: What is this for?

**Comment by thinkingmedia** [comment]
Sorry guys, I was switching branches and forgot this branch was connected to the PR.

I'll clean it up later night.

File: src/Database/Dialect/PostgresDialectTrait.php (hunk 1 above)
Comment (thinkingmedia) [inline src/Database/Dialect/PostgresDialectTrait.php]: @lorenzo oh, okay. I'm just paranoid and don't trust DB engines. I'll remove it.

File: src/Database/Dialect/PostgresDialectTrait.php (hunk 3 above)
Comment (thinkingmedia) [inline src/Database/Dialect/PostgresDialectTrait.php]: I copy/paste so I had an example. It's my first translator :)

File: src/Database/Driver/Mysql.php (hunk 4 above)
Comment (thinkingmedia) [inline src/Database/Driver/Mysql.php]: fixed.

File: src/Database/Dialect/PostgresDialectTrait.php (hunk 5 above)
Comment (thinkingmedia) [inline src/Database/Dialect/PostgresDialectTrait.php]: should not have been in the PR. sorry.

**Comment by thinkingmedia** [comment]
I'm using OID to identify rows in the UPDATE query for Postgres, but these are **optional** on user tables. I'm open to other solutions.

https://www.postgresql.org/docs/8.1/static/datatype-oid.html

File: src/Database/Dialect/PostgresDialectTrait.php [hunk 6]
```diff
@@ -76,15 +74,15 @@ protected function _updateQueryTranslator($query)
 
         if ($limit) {
             $from = (new Query($query->connection()))
-                ->select('*')
+                ->select('oid')
```
Comment (lorenzo) [inline src/Database/Dialect/PostgresDialectTrait.php]: Not all tables have oid

Comment (thinkingmedia) [inline src/Database/Dialect/PostgresDialectTrait.php]: @lorenzo yeah, I know. See my comment above.

**Comment by thinkingmedia** [comment]
I think I need to add the primary keys to the Query object so it can be used to translate queries. I don't see a work around without it.

File: src/Database/Dialect/PostgresDialectTrait.php [hunk 7]
```diff
@@ -76,15 +74,15 @@ protected function _updateQueryTranslator($query)
...
                 ->where($query->clause('where'))
-                ->limit($limit)
-                ->epilog('FOR UPDATE');
+                ->limit($limit);
 
             $outer = clone $query;
             $outer
-                ->from(['__cake_paging__' => $from], true)
+                ->from(['__cake_update__' => $from], true)
+                ->where(['oid'=>$query->newExpr('__cake_update__.oid')], [], true)
```
Comment (ADmad) [inline src/Database/Dialect/PostgresDialectTrait.php]: Space missing around `=>`

**Comment by lorenzo** [comment]
@thinkingmedia as I said, I can do that part, just give me a little time

**Comment by thinkingmedia** [comment]
@lorenzo sure no problem, I'm just going to finish up what my current idea is and then let you review it. You can throw out anything you think is wrong.

**Comment by lorenzo** [comment]
@thinkingmedia If you want to advance your idea a bit further, let me suggest that you throw an exception if the update has a limit, but not an order. A query without an order will make no sense in sql server or postgres, and it is also kind of dangerous in mysql

**Comment by thinkingmedia** [comment]
@lorenzo I was going to add a way to extract the primary keys from the query object, but ran into the problem that the unit tests do not use the ORM version of Query.

```
    /**
     * Returns the primary key associated with the repository that created the query.
     *
     * @param \Cake\Database\Query $query The query to read the primary key from.
     * @return array|null The primary key or null if not found.
     */
    protected function _primaryKey($query)
    {
        if ($query instanceof \Cake\ORM\Query) {
            $table = $query->repository();
            if ($table instanceof \Cake\ORM\Table) {
                return (array)$table->primaryKey();
            }
        }

        return null;
    }
```

The above feels a little hacky, but without the primary keys. I see no way of doing a subquery in either Protgres or SQLite, because they both have the same problem.

The `FROM` clause in Postgres is just a JOIN. So you still need to have a well defined WHERE clause. Otherwise the `FROM` clause does nothing to define the range for the update. As in, it will not be limited. At least, that's my understanding.

**Comment by thinkingmedia** [comment]
@lorenzo SQLite supports ORDER and LIMIT for UPDATE queries. :+1: 

https://www.sqlite.org/lang_update.html

So I'll just add unit tests for the syntax.

I'm going to push my changes. @lorenzo  I'm giving up on Postgres as I've never used the DB. I'll follow your lead on those changes :) No pressure, please take your time.

**Comment by lorenzo** [comment]
@thinkingmedia I pushed some changes to a new branch. I think you can finish my work on top of that https://github.com/cakephp/cakephp/compare/update-limit

**Comment by thinkingmedia** [comment]
@lorenzo thank you, I'll merge that in shortly.

**Comment by thinkingmedia** [comment]
I'll be getting back to this soon. sorry for the delay.

**Comment by lorenzo** [comment]
@thinkingmedia Keep in mind that ALL drivers will produced different SQL, and that the SQL depends on the `quoteIdentifiers` flag. The less you abstract the test cases for testing the generated SQL, the easier it will be for all :)

**Comment by thinkingmedia** [comment]
@lorenzo true, but upon reviewing the unit tests I found that only the edge cases were being tested. I made those changes a while ago. So I'm not exactly sure where I was headed. I think my feeling was to document in the tests what we expect all drivers to provide.

Every time I want to make a PR I have to push to this branch and switch. I might close this PR until I have something ready to be reviewed.

**Comment by thinkingmedia** [comment]
I wish I had not left this sit for so long. Have to figure out where I was and what was left to finish.

File: src/Database/PostgresCompiler.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.3.2
```
Comment (markstory) [inline src/Database/PostgresCompiler.php]: 3.4.0 might be a safer target.

File: tests/TestCase/Database/Driver/PostgresTest.php [hunk 9]
```diff
@@ -171,4 +173,36 @@ public function testInsertReturning()
...
+            ->will($this->returnValue($driver));
+
+        $query = new \Cake\Database\Query($connection);
+
+        $query->update('articles')
+            ->set(['title' => 'FooBar'])
+            ->where(['published' => true])
+            ->limit(5);
+
+        $this->assertEquals('UPDATE articles SET title = :c0 FROM (SELECT oid FROM articles WHERE published = :c1 LIMIT 5) __cake_update__ WHERE oid = (__cake_update__.oid)', $query->sql());
```
Comment (markstory) [inline tests/TestCase/Database/Driver/PostgresTest.php]: Do all tables have `oid`?

Comment (thinkingmedia) [inline tests/TestCase/Database/Driver/PostgresTest.php]: I think this code is going to be removed. I'm going to merge in some changes from @lorenzo but the only way to be sure is to run tests again Postgres and not just verify the generated SQL.

Comment (lorenzo) [inline tests/TestCase/Database/Driver/PostgresTest.php]: No, they don't. In my branch I used "cid" which all tables have

**Comment by thinkingmedia** [comment]
@lorenzo what do you think?

Your code fix is awesome. thank you.

Right now it'll only use a sub-query if a limit is set, but shouldn't this also be done if order is set?

**Comment by thinkingmedia** [comment]
@markstory do you think this should be changed to `3.next`?

**Comment by lorenzo** [comment]
@thinkingmedia The order without a limit has no meaning, so we don't need a subquery in that case

File: src/Database/Dialect/PostgresDialectTrait.php [hunk 10]
```diff
@@ -62,6 +63,53 @@ protected function _transformDistinct($query)
     }
 
     /**
+     * Returns the primary key associated with the repository that created the query.
+     *
+     * @param \Cake\Database\Query $query The query to read the primary key from.
+     * @return array|null The primary key or null if not found.
+     */
+    protected function _primaryKey($query)
```
Comment (lorenzo) [inline src/Database/Dialect/PostgresDialectTrait.php]: What is this for?

Comment (thinkingmedia) [inline src/Database/Dialect/PostgresDialectTrait.php]: That's no longer used and will be removed.

**Comment by markstory** [comment]
> ... do you think this should be changed to 3.next?

Would be the more conservative approach.

**Comment by thinkingmedia** [comment]
Open new PR to fix rebasing on `3.next` here #9730

**Pull request closed without merging**